### PR TITLE
MRG: Add rtmixer sound playback

### PR DESCRIPTION
**Pull request by larsoner**
There are still many TODOs, the biggest of which I can think of now are:

- [x] `_init_mixer` should probably use the preferred *sound card* rather than *given sound stimulus* settings to initialize a sound card (`n_channels, fs`), then we should handle upmixing (mono->stereo if necessary) and software resampling to get the desired rate, rather than trying to always set up a device with `n_channels` and the given `fs`. This is closer to what we do with the TDT, and probably what we want to do to be safest.
- [x] Actually document libs (portaudio windows from spatialaudio; windows rtmixer wheel) and procedures for setup?
- [x] Upload Windows wheels for `python-rtmixer` to facilitate installations
- [x] The CIs break because `rtmixer` needs an actual sound card.
  - [x] Travis OSX: Install virtual sound card
  - [x] Travis Linux: Works already with pulseaudio
  - [x] Appveyor: install vbcable driver (download, extract, install cert, run devcon)
  - [x] CircleCI Linux: pulseaudio currently does not run, stick with Pyglet there for now
- [x] Add Mesa OpenGL rendering to AppVeyor so it actually does visual tests
- [x] Add LGTM analysis
- [x] Fix a few LGTM bugs

Todo separately:

- Actually test timing/latency/jitter
- Get "start sound in N seconds" support working, tested, and integrated in `start_stimulus(when)` to hopefully make A/V timing reliable

But at least at this point I was able to play some basic stimuli. I started on OSX, got things working, then moved to windows, got things moving there, then back to OSX, and most recently, Linux. Hopefully it still works properly on all three.

@drammock I'll try to take care of at least the first bullet point before next week

### CODE REVIEW

**Comment by rkmaddox** [comment]
> Get "start sound in N seconds" support working, tested, and integrated in start_stimulus(when) to hopefully make A/V timing reliable

If the latency is reliable, this should be as simple as zero-padding the beginning of the sound.

Agreed that there should be stream settings (I think that is what you mean by sound card settings) and we can resample on the fly if need be.

So if I were to install the branch now, would I possibly be able to use my RME with `rt_mixer`?

**Comment by rkmaddox** [comment]
One other thing to note that may be obvious--need to be able to select any soundcard, especially ones of which windows is not "aware." By that I mean that in the RME settings you can choose how many channels (actually sets of two channels making stereo pairs) to expose as WDM soundcards. Any number more than 0 makes the RME crash when accessing it through python using `sounddevice` and I imagien `rt_mixer` would have the same problem.

**Comment by larsoner** [comment]
> If the latency is reliable

It won't be (at least A/V won't be) -- this "start in N seconds" is how it should / will become reliable.

> So if I were to install the branch now, would I possibly be able to use my RME with rt_mixer?

You can try, it might work.

> One other thing to note that may be obvious--need to be able to select any soundcard, especially ones of which windows is not "aware." By that I mean that in the RME settings you can choose how many channels (actually sets of two channels making stereo pairs) to expose as WDM soundcards. Any number more than 0 makes the RME crash when accessing it through python using sounddevice and I imagien rt_mixer would have the same problem.

We need a way to access the hardware, and Windows gives us that access. So we'll have to fix `sounddevice` to deal with this properly. `rtmixer` builds directly on top of `sounddevice` (and uses `sounddevice` to query cards and APIs).

**Comment by rkmaddox** [comment]
`sounddevice` already allows using soundcards that the WDM doesn't see. So we should be good there. It's mostly just making sure `expyfun` allows the same selection.

For reliable latency, I thought that was the main point of `rtmixer`, other than having a process that is outside the GIL. The front page of the repo says that those are goals, but also notes they are still to be verified:

>- fixed latency playback, (close to) no jitter (optional)
>    - to be verified ...
>- sample-accurate playback/recording (with known offset)
>    - to be verified ...

We ought to be able to use a system with a TDT also hooked up to do this verification.

**Comment by larsoner** [comment]
IIUC the "fixed latency playback" is achieved by saying "play this in N seconds" where N is greater than your sound card buffer duration

**Comment by larsoner** [comment]
(i.e., saying "play this immediately" will not yield fixed latency playback, it will play the sound ASAP with some variable latency to now depending on the sound card state)

**Comment by rkmaddox** [comment]
> IIUC the "fixed latency playback" is achieved by saying "play this in N seconds" where N is greater than your sound card buffer duration

That is how I understand it too. Or if you say "play this" without saying N seconds, it plays it at a fixed latency, probably 2*buffer. But if they have already done that, then all we should have to do is zero-pad to N - 2*buffer.

**Comment by rkmaddox** [comment]
>(i.e., saying "play this immediately" will not yield fixed latency playback, it will play the sound ASAP with some variable latency to now depending on the sound card state)

Maybe that is where the "optional" comes in for fixed latency.

**Comment by drammock** [comment]
some FYIs from testing on Linux (with the caveat that Linux is nobody's target system at the moment):

1. needed to install `libffi-dev` and `portaudio19-dev`
2. even after that, `python-rtmixer` won't compile, because `gcc: error: portaudio/src/common/pa_ringbuffer.c: No such file or directory`

**Comment by rkmaddox** [comment]
I think that is what is going on [here](https://github.com/spatialaudio/python-rtmixer/blob/d7095f228f58d08fa69916d3657e4996526144cd/src/rtmixer.c#L139). But I will stop distracting you now.

**Comment by larsoner** [comment]
@drammock you need `git submodule init; git submodule update` then it will compile

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** and **fixes 5** when merging 1eab2cded7a73119a3c66ac483fafa2a446e775e into cee010699a4526e215aa0ff21e7420d806307b4a - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-53514e7ea86e976d776c7407b85feb42b0ad05a9)

**new alerts:**

* 1 for \_\_del\_\_ is called explicitly

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **fixes 5 alerts** when merging e54db04d7e1a473ac71452541a0554ad4df541af into cee010699a4526e215aa0ff21e7420d806307b4a - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-59802e8e0e62443891e94883bcc3c46cd58273f1)

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by drammock** [comment]
@larsoner I've got this branch working on my linux laptop now (tried "simple experiment", "basic experiment", and "level test"), although I get `Aborted (core dumped)` whenever I `ctrl-d` after an experiment is over. Below is a start at some "setup procedures" text for linux, describing what I had to do:

On Linux, as of April 2019 the conda version of portaudio is flawed, so the following workaround is necessary to get python-sounddevice to correctly detect audio devices in a conda-managed environment:
```
sudo apt install libportaudio2  # probably it's already installed
conda install python-sounddevice future  # rtmixer prerequisites
conda uninstall --force portaudio
pip install rtmixer  # or clone the git repo and do `pip install -e .` from inside it
```

**Comment by larsoner** [comment]
Pushed some commits to modernize our docs and add a getting started section. The PR review of this is going to be bad, sorry :(

**Comment by larsoner** [comment]
This pull request **fixes 5 alerts** when merging 7f36c7588e11606c3d3a23e1f85c03e1ec4b1d21 into 28a21cafd7635bb442815cf38e7cb4feb0726184 - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-c65297d1c5492083a046ed66f84194f496823f14)

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **fixes 5 alerts** when merging b30d002c0554e7e55d7838e4628ffbe96220171b into 28a21cafd7635bb442815cf38e7cb4feb0726184 - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-c0cfa64c328f6f31a0e5b7fd883bf3a35f5c2b15)

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
Rendered doc about `rtmixer` for now:

https://482-11614571-gh.circle-artifacts.com/0/html/rtmixer_installation.html#rtmixer-installation

**Comment by larsoner** [comment]
Okay @drammock this is finally ready for review/merge.

**Comment by larsoner** [comment]
This pull request **fixes 5 alerts** when merging 1f0074b8ac8b654620bb280f85c9bd718a5f44df into 28a21cafd7635bb442815cf38e7cb4feb0726184 - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-52abd9a1c2c5b6928b2ccd254b349a4d6b92393b)

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **fixes 5 alerts** when merging f7ae27d008bb4b6052dbf1abde897a60ab9f44d5 into dff028e01985506c3077d849eb46c7c7eed36ba2 - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-35e931bf410b6ca944c25d662d527f073ef5a380)

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
Latest docs for ease of perusal

https://491-11614571-gh.circle-artifacts.com/0/html/index.html

**Comment by larsoner** [comment]
This pull request **fixes 5 alerts** when merging 0cb047e0b8d69ce57e496d28d5003b40b54e5e76 into dff028e01985506c3077d849eb46c7c7eed36ba2 - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-7467461a3aaa1fd51888e1667cd5b9ac77419959)

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **fixes 5 alerts** when merging 8e7524d076f3c01dc7a60822e38dfb3f78c7c94a into dff028e01985506c3077d849eb46c7c7eed36ba2 - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-4625ae063b69fd56e8ddd6e255b1d992eddac641)

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: make/install_opengl.ps1 [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        $webclient.DownloadFile($url, $filepath)
+    }
+}
+
+
+function main () {
+    DownloadMesaOpenGL "64"
+}
+
+main
```
Comment (drammock) [inline make/install_opengl.ps1]: missing newline at EOF

File: doc/_static/style.css [hunk 2]
```diff
@@ -13,3 +13,7 @@ div.highlight-console div.highlight span.gp {
     max-width: unset !important;
     min-width: unset !important;
 }
+/* Global TOC */
+li.globaltoc-container {
+   visibility: hidden;
+}
```
Comment (drammock) [inline doc/_static/style.css]: missing newline at EOF

**Comment by drammock** [comment]
@larsoner `make test-doc` is failing for me:

```
SKIP [1] /opt/miniconda3/envs/mnedev/lib/python3.6/site-packages/_pytest/doctest.py:361: unable to import module local('/opt/expyfun/doc/git_diagram.py')
```

Otherwise, besides my comments in the review, LGTM.

File: expyfun/_sound_controllers/_rtmixer.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    apis = query_hostapis()
+    # API
+    if api is None:
+        api = get_config('SOUND_CARD_API', None)
+    if api is None:
+        # Eventually we should maybe allow 'Windows WDM-KS',
+        # 'Windows DirectSound', or 'MME'
+        api = dict(
+            darwin='Core Audio',
+            win32='Windows WASAPI',
```
[the remark this answers is not in the document]
Comment (larsoner) [reply expyfun/_sound_controllers/_rtmixer.py]: Standard sound cards should all support WASAPI.

We could think of eventually adding a try/except to try ASIO first but it's probably not worth the effort. On low-latency experimental machines the config values should always be set anyway

File: expyfun/_sound_controllers/_rtmixer.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                % (param_str, 1000 * device['default_low_output_latency']))
+    return mixer
+
+
+class SoundPlayer(object):
+    """SoundPlayer based on rtmixer."""
+
+    def __init__(self, data, fs=None, loop=False, api=None, name=None):
+        self._data = np.ascontiguousarray(
+            np.clip(data.T, -1, 1).astype(np.float32))
```
[the remark this answers is not in the document]
Comment (larsoner) [reply expyfun/_sound_controllers/_rtmixer.py]: In theory this should really already be taken care of by the callers, this is hopefully just paranoia

File: expyfun/_sound_controllers/_sound_controller.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                'SOUND_CARD_NAME', 'SOUND_CARD_FS')
+        defaults = dict(SOUND_CARD_BACKEND='auto')
+        params = _check_params(params, keys, defaults, 'params')
+
+        self.backend, self.backend_name = _import_backend(
+            params['SOUND_CARD_BACKEND'])
+        logger.info('Expyfun: Setting up sound card audio using %s backend'
+                    % (self.backend_name,))
+        self._kwargs = dict(
+            # XXX add more kwarg translations here
```
[the remark this answers is not in the document]
Comment (larsoner) [reply expyfun/_sound_controllers/_sound_controller.py]: forgot to remove

File: expyfun/_sound_controllers/_sound_controller.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        temp_sound = np.zeros((2, 10))
+        temp_sound = self.backend.SoundPlayer(temp_sound, **self._kwargs)
+        self.fs = temp_sound.fs
+        temp_sound.stop()
+        temp_sound.delete()
+        del temp_sound
+
+        # Need to generate at RMS=1 to match TDT circuit
+        n_samples = 2 ** int(np.ceil(np.log2(self.fs * 15.)))
+        noise = np.random.normal(0, 1.0, n_samples)  # 15 secs
```
[the remark this answers is not in the document]
Comment (larsoner) [reply expyfun/_sound_controllers/_sound_controller.py]: The duration does not really matter for the sound of the noise as long as it's sufficiently long. OTOH the ring buffer needs to be a power of two

**Comment by larsoner** [comment]
`make test-doc` might require some work, I'll open a new issue for that. Otherwise I think I've addressed the existing things, so I'll merge once CIs come back happy and we can iterate in less monolithic PRs. Thanks for reviewing this monster

**Comment by larsoner** [comment]
This pull request **fixes 5 alerts** when merging 517321f0bfbf0af5a4822487b6b461c81948de7a into dff028e01985506c3077d849eb46c7c7eed36ba2 - [view on LGTM.com](https://lgtm.com/projects/g/LABSN/expyfun/rev/pr-154f5f7e2ab1f4dabc2a781aecea38a43f1b3bba)

**fixed alerts:**

* 3 for Variable defined multiple times
* 1 for Unnecessary pass
* 1 for Suspicious unused loop iteration variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rkmaddox** [comment]
THANK YOU both. This is a huge upgrade.

**Pull request merged**